### ISSUE TITLE
Process stuck at 100% usage of one CPU core, h2 receiving GoAway frames

### ISSUE THREAD

**Issue by fnune** [hyperium/hyper#3140]
### Version

hyper: 0.14.24
h2: 0.3.15

_Note: in the debugging notes below, different versions may show. But these are the latest versions with which I've reproduced the problem._

### Platform

```
Linux feanor 5.10.0-21-amd64 #1 SMP Debian 5.10.162-1 (2023-01-21) x86_64 GNU/Linux
```

### Description

In my app I do pretty vanilla usage of https://github.com/Byron/google-apis-rs. In my program, I make a request every 10 seconds. **It works just fine for about 15-20 minutes**, but then the process gets stuck at 100% usage of one CPU core, and a response never comes.

```rust
block_on(
    self.hub
        .events()
        .list("primary")
        .add_scope(Scope::EventReadonly)
        .doit(), // here!
)
.map(|(response, events)| {
    // this scope is never reached
})
.map_err(|_| EventsProviderError {})
```

The expected behavior would be that this fails at some point if the response doesn't return, but it never fails (well, I've left it running for up to 40 minutes, multiple times).

Under the hood, `google-calendar-3` uses hyper:

```rust
let mut req_result = {
    let client = &self.hub.client;
    dlg.pre_request();
    let mut req_builder = hyper::Request::builder().method(hyper::Method::GET).uri(url.clone().into_string())
            .header(USER_AGENT, self.hub._user_agent.clone())                            .header(AUTHORIZATION, format!("Bearer {}", token.as_str()));


            let request = req_builder
            .body(hyper::body::Body::empty());

    client.request(request.unwrap()).await
};
```

Interesting debug logs are these `GoAway` frames and `session_timed_out` debug data, which happen (reproducibly) 4 minutes after the process gets stuck:

```
2023-02-03T23:35:09.720770Z DEBUG Connection{peer=Client}: h2::codec::framed_read: received frame=GoAway { error_code: NO_ERROR, last_stream_id: StreamId(2147483647), debug_data: b"session_timed_out" }
2023-02-03T23:35:09.720832Z DEBUG Connection{peer=Client}: h2::codec::framed_read: received frame=Ping { ack: false, payload: [0, 0, 0, 0, 0, 0, 0, 220] }
2023-02-03T23:35:09.720853Z DEBUG Connection{peer=Client}: h2::codec::framed_write: send frame=Ping { ack: true, payload: [0, 0, 0, 0, 0, 0, 0, 220] }
2023-02-03T23:35:09.720941Z DEBUG Connection{peer=Client}: h2::codec::framed_write: send frame=GoAway { error_code: NO_ERROR, last_stream_id: StreamId(0) }
2023-02-03T23:35:09.720962Z DEBUG Connection{peer=Client}: h2::proto::connection: Connection::poll; connection error error=GoAway(b"", NO_ERROR, Library)
```

My suspicion is that hyper is receiving an response it's not handling. I don't have a mechanism to add a timeout that works, so my program is in an irrecoverable state.

## Debugging notes

Sorry if this is too much for an issue. I've been trying my best to figure out what's going on.

### Basic print debugging

I vendored `google-calendar-3` and added some basic print debugging. Looks like the response just never comes back and the program doesn't go further than the await here:

![image](https://user-images.githubusercontent.com/16181067/216764330-df68c2a6-6440-497b-83e0-61e78a95e8f9.png)

I tried adding a timeout around this and it never times out (same behavior):

```rust
timeout(Duration::from_secs(10), client.request(request.unwrap())).await.unwrap()
```

### Process status when the hang happens

When this happens, the process goes from `interruptible sleep` to `running` (`Sl+` to `Rl+`):

```
fausto    408829  0.5  0.1 967872 37708 pts/2    Sl+  23:49   0:01 ./target/debug/joinable
 ~/Development/joinable => ps aux | grep 408829 | grep -v grep
fausto    408829  0.5  0.1 967872 37708 pts/2    Sl+  23:49   0:01 ./target/debug/joinable
 ~/Development/joinable => ps aux | grep 408829 | grep -v grep
fausto    408829  4.1  0.1 967872 37708 pts/2    Rl+  23:49   0:15 ./target/debug/joinable
```

So maybe it's not waiting for IO anymore.

### GDB backtrace

Got a backtrace from gdb. Looks like the process is stuck between these two and goes into an infinite loop there:

```
futures-executor-0.3.25/src/local_pool.rs:95
std/src/thread/local.rs:446
```

_Note: I got this backtrace by attaching to the running process. I'm not completely sure whether I got the right frames, but this seemed like the most promising output to me._

<details>
  <summary>Full backtrace (expand)</summary>
  <pre>
(gdb) backtrace

#0 futures_executor::local_pool::run_executor::{closure#0}<core::result::Result<(http::response::Response<hyper::body::body::Body>,
google_calendar3::api::Events), google_calendar3::client::Error>,
futures_executor::local_pool::block_on::{closure_env#0}<core::future::from_generator::GenFuture<google_calendar3::api::{impl#133}::doit::{async_fn_env#0}<hyper_rustls::connector::HttpsConnector<hyper::client::connect::http::HttpConnector<hyper::client::connect::dns::GaiResolver>>>>>>
(thread_notify=0x7f1e0b9afc30) at
/home/fausto/.cargo/registry/src/github.com-1ecc6299db9ec823/futures-executor-0.3.25/src/local_pool.rs:95

#1  0x000055b4bc25bbb1 in
std::thread::local::LocalKey<alloc::sync::Arc<futures_executor::local_pool::ThreadNotify>>::try_with<alloc::sync::Arc<futures_executor::local_pool::ThreadNotify>,
futures_executor::local_pool::run_executor::{closure_env#0}<core::result::Result<(http::response::Response<hyper::body::body::Body>,
google_calendar3::api::Events), google_calendar3::client::Error>,
futures_executor::local_pool::block_on::{closure_env#0}<core::future::from_generator::GenFuture<google_calendar3::api::{impl#133}::doit::{async_fn_env#0}<hyper_rustls::connector::HttpsConnector<hyper::client::connect::http::HttpConnector<hyper::client::connect::dns::GaiResolver>>>>>>,
core::result::Result<(http::response::Response<hyper::body::body::Body>,
google_calendar3::api::Events), google_calendar3::client::Error>>
(self=0x55b4bcf505a8, f=...) at
/rustc/69f9c33d71c871fc16ac445211281c6e7a340943/library/std/src/thread/local.rs:446

#2  0x000055b4bc25b3aa in
std::thread::local::LocalKey<alloc::sync::Arc<futures_executor::local_pool::ThreadNotify>>::with<alloc::sync::Arc<futures_executor::local_pool::ThreadNotify>,
futures_executor::local_pool::run_executor::{closure_env#0}<core::result::Result<(http::response::Response<hyper::body::body::Body>,
google_calendar3::api::Events), google_calendar3::client::Error>,
futures_executor::local_pool::block_on::{closure_env#0}<core::future::from_generator::GenFuture<google_calendar3::api::{impl#133}::doit::{async_fn_env#0}<hyper_rustls::connector::HttpsConnector<hyper::client::connect::http::HttpConnector<hyper::client::connect::dns::GaiResolver>>>>>>,
core::result::Result<(http::response::Response<hyper::body::body::Body>,
google_calendar3::api::Events), google_calendar3::client::Error>>
(self=0x55b4bcf505a8, f=...) thread/local.rs:422

#3  0x000055b4bc126bbf in
futures_executor::local_pool::run_executor<core::result::Result<(http::response::Response<hyper::body::body::Body>,
google_calendar3::api::Events), google_calendar3::client::Error>,
futures_executor::local_pool::block_on::{closure_env#0}<core::future::from_generator::GenFuture<google_calendar3::api::{impl#133}::doit::{async_fn_env#0}<hyper_rustls::connector::HttpsConnector<hyper::client::connect::http::HttpConnector<hyper::client::connect::dns::GaiResolver>>>>>>
(f=...) at
/home/fausto/.cargo/registry/src/github.com-1ecc6299db9ec823/futures-executor-0.3.25/src/local_pool.rs:86

#4  0x000055b4bc1270ce in
futures_executor::local_pool::block_on<core::future::from_generator::GenFuture<google_calendar3::api::{impl#133}::doit::{async_fn_env#0}<hyper_rustls::connector::HttpsConnector<hyper::client::connect::http::HttpConnector<hyper::client::connect::dns::GaiResolver>>>>>
(f=...) at
/home/fausto/.cargo/registry/src/github.com-1ecc6299db9ec823/futures-executor-0.3.25/src/local_pool.rs:317

#5  0x000055b4bc091ccf in
joinable::events::providers::google_calendar::{impl#2}::fetch_events
(self=0x7f1df8001478) at src/events/providers/google_calendar.rs:87
  </pre>
</details>

### Wireshark and `tokio-console`

I tried to take a further look at what packets are being passed around:

![image](https://user-images.githubusercontent.com/16181067/216764991-24ff2e6a-1ae2-41c0-80fb-de114de26329.png)

My program was polling every 10 seconds for testing, so (1) and (2) show successful requests and all their packets (the box after 2 is also a successful request).

Section (3) shows the closing of the TCP connection when I realized the program got stuck and no further packets were received (I hit Ctrl+C).

So the finding is: the Rust program is stuck before the response packets arrive.

`tokio-console` output looks like this:

![image](https://user-images.githubusercontent.com/16181067/216765041-e62522a4-0308-4f3a-b7bb-051c6997e87e.png)

This is the last resource that pops up and never stops being used:

```
│   110 n/a     Sync   198.8622s tokio::sync::oneshot         Sender|Receiver ✅   <cargo>/hyper-0.14.23/src/client/dispatch.rs:125:24      rx_dropped=false tx_dropped=false value_received=false value_sent=true              │
```

_Note: I don't understand the output of `tokio-console` completely, as I'm not super experienced with async Rust)._

The resource is this one-shot channel here:

https://github.com/hyperium/hyper/blob/0.14.x/src/client/dispatch.rs#L124-L130

The number of polls in the tasks view of tokio-console is no longer increasing (although that could be because the program is stuck, and is no longer emitting traces).

### New debug logs after I added `tokio-console`

Looks like enabling the `tracing` feature of Tokio made my logs show some more stuff...

This is a healthy fetch:

```
2023-02-03T23:30:59.593351Z DEBUG hyper::client::pool: reuse idle connection for ("https", www.googleapis.com)
2023-02-03T23:30:59.594418Z DEBUG Connection{peer=Client}: h2::codec::framed_write: send frame=Headers { stream_id: StreamId(217), flags: (0x5: END_HEADERS | END_STREAM) }
2023-02-03T23:30:59.726889Z DEBUG Connection{peer=Client}: h2::codec::framed_read: received frame=Headers { stream_id: StreamId(217), flags: (0x4: END_HEADERS) }
2023-02-03T23:30:59.727045Z DEBUG Connection{peer=Client}: h2::codec::framed_read: received frame=Data { stream_id: StreamId(217) }
2023-02-03T23:30:59.727125Z DEBUG Connection{peer=Client}: h2::codec::framed_read: received frame=Data { stream_id: StreamId(217), flags: (0x1: END_STREAM) }
2023-02-03T23:30:59.727204Z DEBUG Connection{peer=Client}: h2::codec::framed_read: received frame=Ping { ack: false, payload: [0, 0, 0, 0, 0, 0, 0, 216] }
2023-02-03T23:30:59.727249Z DEBUG Connection{peer=Client}: h2::codec::framed_write: send frame=Ping { ack: true, payload: [0, 0, 0, 0, 0, 0, 0, 216] }
```

This is the fetch where the process gets stuck:

```
2023-02-03T23:31:09.591212Z DEBUG hyper::client::pool: reuse idle connection for ("https", www.googleapis.com)
2023-02-03T23:31:09.592015Z DEBUG Connection{peer=Client}: h2::codec::framed_write: send frame=Headers { stream_id: StreamId(219), flags: (0x5: END_HEADERS | END_STREAM) }
2023-02-03T23:31:09.718778Z DEBUG Connection{peer=Client}: h2::codec::framed_read: received frame=Headers { stream_id: StreamId(219), flags: (0x4: END_HEADERS) }
2023-02-03T23:31:09.718854Z DEBUG Connection{peer=Client}: h2::codec::framed_read: received frame=Data { stream_id: StreamId(219) }
2023-02-03T23:31:09.719094Z DEBUG Connection{peer=Client}: h2::codec::framed_read: received frame=Data { stream_id: StreamId(219), flags: (0x1: END_STREAM) }
2023-02-03T23:31:09.719142Z DEBUG Connection{peer=Client}: h2::codec::framed_read: received frame=Ping { ack: false, payload: [0, 0, 0, 0, 0, 0, 0, 218] }
2023-02-03T23:31:09.719163Z DEBUG Connection{peer=Client}: h2::codec::framed_write: send frame=Ping { ack: true, payload: [0, 0, 0, 0, 0, 0, 0, 218] }
2023-02-03T23:35:09.720770Z DEBUG Connection{peer=Client}: h2::codec::framed_read: received frame=GoAway { error_code: NO_ERROR, last_stream_id: StreamId(2147483647), debug_data: b"session_timed_out" }
2023-02-03T23:35:09.720832Z DEBUG Connection{peer=Client}: h2::codec::framed_read: received frame=Ping { ack: false, payload: [0, 0, 0, 0, 0, 0, 0, 220] }
2023-02-03T23:35:09.720853Z DEBUG Connection{peer=Client}: h2::codec::framed_write: send frame=Ping { ack: true, payload: [0, 0, 0, 0, 0, 0, 0, 220] }
2023-02-03T23:35:09.720941Z DEBUG Connection{peer=Client}: h2::codec::framed_write: send frame=GoAway { error_code: NO_ERROR, last_stream_id: StreamId(0) }
2023-02-03T23:35:09.720962Z DEBUG Connection{peer=Client}: h2::proto::connection: Connection::poll; connection error error=GoAway(b"", NO_ERROR, Library)
[2023-02-03T23:35:09Z DEBUG rustls::conn] Sending warning alert CloseNotify
```

Particularly this:

![image](https://user-images.githubusercontent.com/16181067/216731074-482286ff-a88f-45c9-87fc-27bbefb4bd53.png)

Note that that first frame with `session_timed_out` debug data happens exactly 4 minutes after the process becomes stuck.

**Comment by fnune** [comment]
Note that these notes are all upstreamed from https://github.com/Byron/google-apis-rs/issues/402

**Comment by fnune** [comment]
This looks related: https://stackoverflow.com/questions/55087292/how-to-handle-http-2-goaway-with-httpclient

So probably my connection has exhausted a server-side per-connection limit and is being closed by the server. Maybe hyper is not handling that?

**Comment by fnune** [comment]
Looks like `golang/net` retries when it gets `GoAway`:

https://github.com/golang/net/commit/8dab9293431a241c65a8e8d46bbaa0492ec94849

**Comment by fnune** [comment]
Trying (but failing) to reproduce this here: https://github.com/fnune/hyper-goaway-repro

I get an expected/welcome immediate failure:

```rust
 ~/Development/hyper-goaway-repro [101] => cargo run --bin client                                      14:09:47
    Finished dev [unoptimized + debuginfo] target(s) in 0.02s
     Running `target/debug/client`
thread 'main' panicked at 'called `Result::unwrap()` on an `Err` value: hyper::Error(Io, Kind(BrokenPipe))', src/client/bin/main.rs:18:26
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
```

What I'm looking for is to emulate the process hanging problem...

**Comment by fnune** [comment]
![image](https://user-images.githubusercontent.com/16181067/216803168-f8fe5cd2-f108-4041-bb13-626fe04aeb14.png)

That's disappointing, ChatGPT...

**Comment by fnune** [comment]
Recorded the GoAway frames with Wireshark:

![image](https://user-images.githubusercontent.com/16181067/216805417-6d487f3d-9199-441d-a38d-fea1eed714e5.png)

(1) The server sends `[FIN, ACK]`
(2) The client ignores the packet and sends `[RST]`

Once more:

![image](https://user-images.githubusercontent.com/16181067/216806191-284128f6-994f-4086-8403-df53c0df4d9e.png)

They happen ~18-20 minutes after the connection begins.

**Comment by seanmonstar** [comment]
> futures-executor-0.3.25/src/local_pool.rs:95

This is a red flag to me. I would check where you are calling the local pool, (is that `block_on` you mention at the top it?), and rip them out. They don't work well with Tokio's reactor. Replace them with either a tokio runtime `block_on` call, or better, make the function async itself, and just use `#[tokio::main]`.

The reason they conflict: Tokio's reactor uses `epoll_wait` when it needs to wait on IO. When epoll returns, Tokio's reactor processes all events it noticed. By using `futures-executor`, _it's_ `block_on` does something a little different. When the future is not ready, it will park the thread (`std::thread::park()`) until the future wakes up. The problem is that the reactor is on that same thread, and since it parked, it will stall there, and never reach the `epoll_wait`. Using `futures-executor` is a very common cause for deadlocks.

**Comment by fnune** [comment]
Oh no! That makes sense...

I suppose the unexpected communication at the TCP level I'm seeing was a red herring, then, and the client isn't responding correctly to the `[FIN, ACK]` simply because my thread is parked?

**Comment by seanmonstar** [comment]
It's hard for me to say exactly. The TCP stuff happens inside the kernel, and isn't necessarily blocked by the user-space thread parking. But it won't ever *wake up* that thread.

I mean it could be something else, but that is the most common culprit if `futures-executor` is involved.

**Comment by fnune** [comment]
I'm trying to fit the 100% core usage within the framework of your (eye-opening) explanation, but I can't figure it out. I suppose while the thread is parked something is polling to see if it needs to be waken up?

(I'll be on my computer trying out your suggested patch in some minutes)

Thank you so much for your time and help already!

**Comment by fnune** [comment]
Okay, I turned the whole code path async to be able to make better use of Tokio like you suggested. I'm no longer using `block_on` (from any source) and things are working perfectly!

```
[2023-02-06T19:35:34Z DEBUG h2::codec::framed_write] send frame=GoAway { error_code: NO_ERROR, last_stream_id: StreamId(0) }
[2023-02-06T19:35:34Z DEBUG h2::proto::connection] Connection::poll; connection error error=GoAway(b"", NO_ERROR, Library)
[2023-02-06T19:35:34Z DEBUG rustls::conn] Sending warning alert CloseNotify
[2023-02-06T19:35:34Z DEBUG hyper::client::connect::dns] resolving host="www.googleapis.com"
[2023-02-06T19:35:34Z DEBUG hyper::client::connect::http] connecting to 142.251.209.138:443
[2023-02-06T19:35:34Z DEBUG hyper::client::connect::http] connected to 142.251.209.138:443
[2023-02-06T19:35:34Z DEBUG rustls::client::hs] No cached session for DnsName(DnsName(DnsName("www.googleapis.com")))
[2023-02-06T19:35:34Z DEBUG rustls::client::hs] Not resuming any session
[2023-02-06T19:35:34Z DEBUG rustls::client::hs] Using ciphersuite TLS13_AES_256_GCM_SHA384
[2023-02-06T19:35:34Z DEBUG rustls::client::tls13] Not resuming
[2023-02-06T19:35:34Z DEBUG rustls::client::tls13] TLS1.3 encrypted extensions: [Protocols([6832])]
[2023-02-06T19:35:34Z DEBUG rustls::client::hs] ALPN protocol is Some(b"h2")
[2023-02-06T19:35:34Z DEBUG h2::client] binding client connection
[2023-02-06T19:35:34Z DEBUG h2::client] client connection bound
[2023-02-06T19:35:34Z DEBUG h2::codec::framed_write] send frame=Settings { flags: (0x0), enable_push: 0, initial_window_size: 2097152, max_frame_size: 16384 }
[2023-02-06T19:35:34Z DEBUG h2::proto::connection] Connection; peer=Client
[2023-02-06T19:35:34Z DEBUG hyper::client::pool] pooling idle connection for ("https", www.googleapis.com)
[2023-02-06T19:35:34Z DEBUG h2::codec::framed_write] send frame=WindowUpdate { stream_id: StreamId(0), size_increment: 5177345 }
[2023-02-06T19:35:34Z DEBUG h2::codec::framed_write] send frame=Headers { stream_id: StreamId(1), flags: (0x5: END_HEADERS | END_STREAM) }
[2023-02-06T19:35:34Z DEBUG rustls::client::tls13] Ticket saved
[2023-02-06T19:35:34Z DEBUG rustls::client::tls13] Ticket saved
[2023-02-06T19:35:34Z DEBUG h2::codec::framed_read] received frame=Settings { flags: (0x0), max_concurrent_streams: 100, initial_window_size: 1048576, max_header_list_size: 65536 }
[2023-02-06T19:35:34Z DEBUG h2::codec::framed_write] send frame=Settings { flags: (0x1: ACK) }
[2023-02-06T19:35:34Z DEBUG h2::codec::framed_read] received frame=WindowUpdate { stream_id: StreamId(0), size_increment: 983041 }
[2023-02-06T19:35:34Z DEBUG h2::codec::framed_read] received frame=Settings { flags: (0x1: ACK) }
[2023-02-06T19:35:34Z DEBUG h2::proto::settings] received settings ACK; applying Settings { flags: (0x0), enable_push: 0, initial_window_size: 2097152, max_frame_size: 16384 }
[2023-02-06T19:35:34Z DEBUG h2::codec::framed_read] received frame=Headers { stream_id: StreamId(1), flags: (0x4: END_HEADERS) }
[2023-02-06T19:35:34Z DEBUG h2::codec::framed_read] received frame=Data { stream_id: StreamId(1) }
[2023-02-06T19:35:34Z DEBUG h2::codec::framed_read] received frame=Data { stream_id: StreamId(1), flags: (0x1: END_STREAM) }
[2023-02-06T19:35:34Z DEBUG h2::codec::framed_read] received frame=Ping { ack: false, payload: [0, 0, 0, 0, 0, 0, 0, 0] }
[2023-02-06T19:35:34Z DEBUG h2::codec::framed_write] send frame=Ping { ack: true, payload: [0, 0, 0, 0, 0, 0, 0, 0] }
```

For some reason I no longer get the `GoAway` frames after 20 minutes. In the instance above I got them just after first running the program. Makes me wonder at what point the thread got parked. I may continue investigating with a breakpoint on `std::thread::park` I guess.

Thank you so much for your help! You not only fixed my program but also helped me realize that I need to learn more about reactors and async Rust in general.

**Comment by seanmonstar** [comment]
Nice! I'm glad it was that simple.

(It would be nice if Tokio could somehow detect this, but most obvious ways have other downfalls. :()
